src/sync_user.hpp

```diff
@@ -0,0 +1,118 @@
+////////////////////////////////////////////////////////////////////////////
+//
+// Copyright 2016 Realm Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+// http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+//
+////////////////////////////////////////////////////////////////////////////
+
+#ifndef REALM_OS_SYNC_USER_HPP
+#define REALM_OS_SYNC_USER_HPP
+
+#include <string>
+#include <memory>
+#include <unordered_map>
+
+namespace realm {
+
+struct SyncSession;
+
+// A `SyncUser` represents a single user account. Each user manages the sessions that
+// are associated with it.
+class SyncUser {
+public:	
+	enum class State {
+		LoggedOut, Active, Error;
```

Comment:
The session error handler is called with one of the error codes that indicates a user-level (as opposed to session level) error. I believe `Error::bad_authentication` is the only error right now that can cause this issue (i.e. auth and sync servers are misconfigured so the sync server doesn't accept the token issued by the auth server.)

Currently, we treat this as a 'fatal'/'unrecoverable' error that requires the app to try making a different user.